homeassistant/components/bluetooth/base_scanner.py

```diff
@@ -65,6 +71,47 @@ def __init__(
         self.name = adapter_human_name(adapter, source) if adapter != source else source
         self.scanning = True
         self._last_detection = 0.0
+        self._start_time = 0.0
+        self._cancel_watchdog: CALLBACK_TYPE | None = None
+
+    @hass_callback
+    def _async_stop_scanner_watchdog(self) -> None:
+        """Stop the scanner watchdog."""
+        if self._cancel_watchdog:
+            self._cancel_watchdog()
+            self._cancel_watchdog = None
+
+    @hass_callback
+    def _async_setup_scanner_watchdog(self) -> None:
+        """If something has restarted or updated, we need to restart the scanner."""
+        self._start_time = self._last_detection = MONOTONIC_TIME()
+        if not self._cancel_watchdog:
+            self._cancel_watchdog = async_track_time_interval(
+                self.hass, self._async_scanner_watchdog, SCANNER_WATCHDOG_INTERVAL
+            )
+
+    @hass_callback
+    def _async_watchdog_triggered(self) -> bool:
+        """Check if the watchdog has been triggered."""
+        time_since_last_detection = MONOTONIC_TIME() - self._last_detection
+        _LOGGER.debug(
+            "%s: Scanner watchdog time_since_last_detection: %s",
+            self.name,
+            time_since_last_detection,
+        )
+        return time_since_last_detection > SCANNER_WATCHDOG_TIMEOUT
```

Comment:
For esphome and shelly we expect this will never happen since as soon as the device is detected as unavailable the scanner should be unregistered, however there are obviously some cases where that doesn't happen as expected like https://github.com/home-assistant-libs/aioshelly/pull/307 so we now have this safeguard in place so we can get a better idea of whats going on when users report issues